packages/mobx/src/core/derivation.ts

```diff
@@ -166,10 +166,13 @@ export function checkIfStateReadsAreAllowed(observable: IObservable) {
  */
 export function trackDerivedFunction<T>(derivation: IDerivation, f: () => T, context: any) {
     const prevAllowStateReads = allowStateReadsStart(true)
-    // pre allocate array allocation + room for variation in deps
-    // array will be trimmed by bindDependencies
     changeDependenciesStateTo0(derivation)
-    derivation.newObserving_ = new Array(derivation.observing_.length + 100)
+    // Preallocate array; will be trimmed by bindDependencies.
+    derivation.newObserving_ = new Array(
+        derivation.observing_.length
+            ? Math.ceil(derivation.observing_.length * 1.2) // 20% extra room for variation in deps
+            : 100 // Establishing initial dependencies, preallocate constant space.
```

Comment:
Is `derivation.observing_.length` a good indicator of "initial"? Ideally we would like to avoid prealocating anything for non-initial runs if there are no observable deps, eg:
```
const Cmp = observer(() => <div>no-observable-deps</div>)
```
Maybe there is an opportunity to bail-out even earlier is this case, dunno.

Reply:
Hmmm, good question! I think we wouldn't re-run `trackDerivedFunction` in case there are no dependencies, right? (Because nothing "inside" would trigger a rerun and external calls would get the cached value.)

Reply:
No, deps can still change based on something not observable like prop/state.
The main point is that zero deps isn't special case - we assume the same thing - the number of deps don't change significantly between runs. Therefore we preallocate array of the same size - in this case an array of zero length. Probably not much else we can do. But we need to somehow differentiate between initial run (preallocate constant space) and non-initial run (preallocate zero). Probably we can just change the condition to: `derivation.runId_ === 0`